File: modules/photo-vision/ios/PhotoVisionAnalyzer.swift
//
//  PhotoVisionAnalyzer.swift
//  Step 2 — iOS Vision 기반 기술적 품질 평가 (순수 분석 로직)
//
//  ⚠️ OOM 방지 원칙:
//   - 원본 전체를 디코딩하지 않고 CGImageSource 썸네일 API로 '축소본'만 메모리에 올린다.
//   - 흔들림/노출 계산은 256px 작은 그레이스케일 버퍼에서만 수행한다.
//   - 호출부(Module)에서 사진 1장마다 autoreleasepool로 감싸 즉시 해제한다.
//

import Foundation
import CoreGraphics
import ImageIO
import Vision

/// 사진 1장 분석 결과 (JS로 그대로 직렬화)
struct PhotoAnalysis {
    var uri: String
    var blurVariance: Double      // 라플라시안 분산. 클수록 선명 (흐림이면 작음)
    var meanLuminance: Double     // 0~1 평균 밝기 (노출 판단용)
    var aestheticsScore: Double   // 0~1 미학 점수. 미지원(iOS<18)/실패 시 -1
    var isUtility: Bool           // 영수증/문서/스크린샷류 여부 (Vision 판정)
    // ─ 의미 분석 (Step 5) ─
    var hasFace: Bool = false     // 인물 포함 여부
    var isSmiling: Bool = false   // 웃는 얼굴 (iOS Vision은 미지원 → 항상 false, Android에서만 채움)
    var isFood: Bool = false      // 음식
    var isLandscape: Bool = false // 풍경/자연
    var isLandmark: Bool = false  // 건축물/랜드마크
    var error: String?

    func toDict() -> [String: Any] {
        return [
            "uri": uri,
            "blurVariance": blurVariance,
            "meanLuminance": meanLuminance,
            "aestheticsScore": aestheticsScore,
            "isUtility": isUtility,
            "hasFace": hasFace,
            "isSmiling": isSmiling,
            "isFood": isFood,
            "isLandscape": isLandscape,
            "isLandmark": isLandmark,
            "error": error as Any
        ]
    }
}

enum PhotoVisionAnalyzer {

    /// 분석용 최대 변(픽셀). 호출부가 이미 썸네일을 넘기더라도 방어적으로 한 번 더 축소.
    private static let maxAnalysisPixel = 512
    /// 흔들림/노출 계산용 그레이스케일 버퍼 한 변 크기.
    private static let grayEdge = 256

    /// 파일 경로(uri)를 받아 품질 지표를 계산한다.
    static func analyze(uri: String) -> [String: Any] {
        var result = PhotoAnalysis(
            uri: uri,
            blurVariance: 0,
            meanLuminance: 0,
            aestheticsScore: -1,
            isUtility: false,
            error: nil
        )

        guard let url = fileURL(from: uri) else {
            result.error = "INVALID_URI"
            return result.toDict()
        }

        // 1) 축소본(CGImage) 로드 — 원본 풀디코딩 회피
        guard let cgImage = downsampledImage(at: url, maxPixel: maxAnalysisPixel) else {
            result.error = "DECODE_FAILED"
            return result.toDict()
        }

        // 2) 그레이스케일 작은 버퍼 → 흔들림/노출 계산
        if let gray = grayscaleBuffer(from: cgImage, edge: grayEdge) {
            result.meanLuminance = meanLuminance(gray.pixels)
            result.blurVariance = laplacianVariance(gray.pixels, width: gray.width, height: gray.height)
        } else {
            result.error = "GRAYSCALE_FAILED"
        }

        // 3) Vision 미학 점수 + 유틸리티(문서) 판정 — iOS 18+
        if #available(iOS 18.0, *) {
            let aesthetics = aestheticsScore(for: cgImage)
            result.aestheticsScore = aesthetics.score
            result.isUtility = aesthetics.isUtility
        }

        // 4) 의미 분석 (Step 5) — 문서가 아닐 때만 (영수증/지도엔 불필요)
        if !result.isUtility {
            result.hasFace = detectFace(in: cgImage)
            let cls = classify(image: cgImage)
            result.isFood = cls.food
            result.isLandscape = cls.landscape
            result.isLandmark = cls.landmark
        }

        return result.toDict()
    }

    // MARK: - 의미 분석

    /// 얼굴 포함 여부 (iOS 13+). 웃음 판정 API는 없어 hasFace만 반환.
    private static func detectFace(in image: CGImage) -> Bool {
        let request = VNDetectFaceRectanglesRequest()
        let handler = VNImageRequestHandler(cgImage: image, options: [:])
        do {
            try handler.perform([request])
            return (request.results?.isEmpty == false)
        } catch {
            return false
        }
    }

    // VNClassifyImageRequest 식별자(taxonomy)를 키워드로 매핑
    private static let foodKeywords = ["food", "meal", "dish", "fruit", "vegetable",
        "dessert", "drink", "beverage", "coffee", "cake", "pizza", "bread"]
    private static let landscapeKeywords = ["outdoor", "landscape", "mountain", "beach",
        "sky", "sea", "ocean", "sunset", "sunrise", "nature", "field", "forest",
        "lake", "river", "cloud", "tree", "valley", "desert", "snow", "waterfall"]
    private static let landmarkKeywords = ["building", "architecture", "tower",
        "skyscraper", "monument", "castle", "temple", "church", "cathedral",
        "bridge", "statue", "palace", "landmark", "structure"]

    /// 이미지 분류로 음식/풍경/랜드마크 추정 (iOS 13+).
    private static func classify(image: CGImage) -> (food: Bool, landscape: Bool, landmark: Bool) {
        let request = VNClassifyImageRequest()
        let handler = VNImageRequestHandler(cgImage: image, options: [:])
        do {
            try handler.perform([request])
            guard let observations = request.results else { return (false, false, false) }
            let labels = observations
                .filter { $0.confidence > 0.6 }
                .map { $0.identifier.lowercased() }

            let hit = { (keys: [String]) -> Bool in
                labels.contains { id in keys.contains { id.contains($0) } }
            }
            return (hit(foodKeywords), hit(landscapeKeywords), hit(landmarkKeywords))
        } catch {
            return (false, false, false)
        }
    }

    // MARK: - 파일 경로 처리

    private static func fileURL(from uri: String) -> URL? {
        if uri.hasPrefix("file://") {
            return URL(string: uri)
        }
        if uri.hasPrefix("ph://") || uri.hasPrefix("assets-library://") {
            // PHAsset 경로는 미지원: 호출부에서 file:// 썸네일을 넘겨야 함
            return nil
        }
        return URL(fileURLWithPath: uri)
    }

    // MARK: - 메모리 효율적 다운샘플 로드

    private static func downsampledImage(at url: URL, maxPixel: Int) -> CGImage? {
        let srcOptions = [kCGImageSourceShouldCache: false] as CFDictionary
        guard let src = CGImageSourceCreateWithURL(url as CFURL, srcOptions) else { return nil }

        let options: [CFString: Any] = [
            kCGImageSourceCreateThumbnailFromImageAlways: true,
            kCGImageSourceShouldCacheImmediately: true,
            kCGImageSourceCreateThumbnailWithTransform: true,
            kCGImageSourceThumbnailMaxPixelSize: maxPixel
        ]
        return CGImageSourceCreateThumbnailAtIndex(src, 0, options as CFDictionary)
    }

    // MARK: - 그레이스케일 버퍼

    private struct GrayBuffer {
        let pixels: [UInt8]
        let width: Int
        let height: Int
    }

    /// CGImage를 한 변 `edge` 이하의 8bit 그레이스케일 버퍼로 변환.
    private static func grayscaleBuffer(from image: CGImage, edge: Int) -> GrayBuffer? {
        let ratio = min(1.0, Double(edge) / Double(max(image.width, image.height)))
        let w = max(1, Int(Double(image.width) * ratio))
        let h = max(1, Int(Double(image.height) * ratio))

        var pixels = [UInt8](repeating: 0, count: w * h)
        let colorSpace = CGColorSpaceCreateDeviceGray()

        guard let ctx = CGContext(
            data: &pixels,
            width: w,
            height: h,
            bitsPerComponent: 8,
            bytesPerRow: w,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.none.rawValue
        ) else { return nil }

        ctx.draw(image, in: CGRect(x: 0, y: 0, width: w, height: h))
        return GrayBuffer(pixels: pixels, width: w, height: h)
    }

    // MARK: - 노출 (평균 밝기 0~1)

    private static func meanLuminance(_ pixels: [UInt8]) -> Double {
        guard !pixels.isEmpty else { return 0 }
        var sum = 0
        for p in pixels { sum += Int(p) }
        return (Double(sum) / Double(pixels.count)) / 255.0
    }

    // MARK: - 흔들림 (라플라시안 분산)

    /// 3x3 라플라시안 커널을 적용한 응답의 분산. 선명할수록 고주파 성분↑ → 분산↑.
    private static func laplacianVariance(_ pixels: [UInt8], width: Int, height: Int) -> Double {
        guard width > 2, height > 2 else { return 0 }

        var responses = [Double]()
        responses.reserveCapacity((width - 2) * (height - 2))

        for y in 1..<(height - 1) {
            for x in 1..<(width - 1) {
                let c = Int(pixels[y * width + x])
                let up = Int(pixels[(y - 1) * width + x])
                let down = Int(pixels[(y + 1) * width + x])
                let left = Int(pixels[y * width + (x - 1)])
                let right = Int(pixels[y * width + (x + 1)])
                // ∇² = 4*center - (상하좌우 합)
                let lap = Double(4 * c - up - down - left - right)
                responses.append(lap)
            }
        }

        let n = Double(responses.count)
        guard n > 0 else { return 0 }
        let mean = responses.reduce(0, +) / n
        var variance = 0.0
        for r in responses { variance += (r - mean) * (r - mean) }
        return variance / n
    }

    // MARK: - Vision 미학 점수 (iOS 18+)

    @available(iOS 18.0, *)
    private static func aestheticsScore(for image: CGImage) -> (score: Double, isUtility: Bool) {
        let request = VNGenerateImageAestheticsScoresRequest()
        let handler = VNImageRequestHandler(cgImage: image, options: [:])
        do {
            try handler.perform([request])
            if let obs = request.results?.first {
                // overallScore: -1.0 ~ 1.0 → 0~1 정규화
                let normalized = (Double(obs.overallScore) + 1.0) / 2.0
                return (normalized, obs.isUtility)
            }
        } catch {
            // 무시: 점수 미산출(-1)로 폴백
        }
        return (-1, false)
    }
}

File: modules/photo-vision/android/src/main/java/expo/modules/photovision/PhotoVisionAnalyzer.kt
package expo.modules.photovision

//
//  Step 3 — Android ML Kit + 비트맵 기반 기술적 품질 평가 (순수 분석 로직)
//
//  iOS(PhotoVisionAnalyzer.swift)와 동일한 출력 스키마를 반환하여
//  JS 레이어가 플랫폼을 구분하지 않도록 한다.
//
//  ⚠️ OOM 방지:
//   - BitmapFactory inSampleSize 로 축소 디코딩(원본 풀디코딩 회피)
//   - 흔들림/노출은 256px 그레이스케일 버퍼에서만 계산
//   - 사용한 Bitmap 은 즉시 recycle()
//

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.label.ImageLabeling
import com.google.mlkit.vision.label.defaults.ImageLabelerOptions
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import java.io.File
import kotlin.math.max
import kotlin.math.min

object PhotoVisionAnalyzer {

    private const val MAX_ANALYSIS_PIXEL = 512   // 분석용 최대 변
    private const val GRAY_EDGE = 256            // 흔들림/노출 계산용 한 변
    private const val UTILITY_TEXT_LEN = 120     // 공백 제외 텍스트가 이 길이 이상이면 문서/스크린샷류로 추정
    private const val SMILE_PROB = 0.6f          // 웃음 확률 임계값
    private const val LABEL_CONFIDENCE = 0.6f    // 이미지 라벨 신뢰도 임계값

    // 이미지 라벨 → 카테고리 매핑 키워드 (소문자 contains 매칭)
    private val FOOD_KEYWORDS = listOf("food", "meal", "dish", "fruit", "vegetable",
        "dessert", "drink", "coffee", "cake", "bread")
    private val LANDSCAPE_KEYWORDS = listOf("mountain", "beach", "sky", "sea", "ocean",
        "sunset", "sunrise", "nature", "forest", "lake", "river", "cloud", "tree",
        "snow", "waterfall", "landscape")
    private val LANDMARK_KEYWORDS = listOf("building", "tower", "skyscraper", "monument",
        "castle", "temple", "church", "bridge", "statue", "palace", "landmark")

    // 검출기는 배치 전체에서 재사용 (장마다 모델 재로딩 방지)
    private val faceDetector by lazy {
        FaceDetection.getClient(
            FaceDetectorOptions.Builder()
                .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL) // 웃음 확률 필요
                .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
                .build()
        )
    }
    private val imageLabeler by lazy {
        ImageLabeling.getClient(ImageLabelerOptions.DEFAULT_OPTIONS)
    }
    private val textRecognizer by lazy {
        TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
    }

    /** file:// 썸네일 경로 1건을 분석하여 iOS와 동일한 키 구조의 Map 반환 */
    fun analyze(uri: String): Map<String, Any?> {
        val result = hashMapOf<String, Any?>(
            "uri" to uri,
            "blurVariance" to 0.0,
            "meanLuminance" to 0.0,
            "aestheticsScore" to -1.0, // Android(ML Kit)에는 미학 점수 API 없음 → 미지원 센티넬
            "isUtility" to false,
            // ─ 의미 분석 (Step 5) ─
            "hasFace" to false,
            "isSmiling" to false,
            "isFood" to false,
            "isLandscape" to false,
            "isLandmark" to false,
            // ─ 확장 신호 (형식 추천, 2026-08-31) ─
            "sceneLabels" to emptyList<Map<String, Any>>(),
            "faceCount" to 0,
            "hasText" to false,
            "colorStats" to emptyMap<String, Double>(),
            "dhash" to "",
            "error" to null
        )

        val path = filePath(uri)
        if (path == null) {
            result["error"] = "INVALID_URI"
            return result
        }

        val bitmap = decodeDownsampled(path, MAX_ANALYSIS_PIXEL)
        if (bitmap == null) {
            result["error"] = "DECODE_FAILED"
            return result
        }

        try {
            val gray = toGray(bitmap, GRAY_EDGE)
            result["meanLuminance"] = meanLuminance(gray.lum)
            result["blurVariance"] = laplacianVariance(gray.lum, gray.width, gray.height)
            result["dhash"] = differenceHash(bitmap)
            result["colorStats"] = colorStatistics(bitmap)

            val image = InputImage.fromBitmap(bitmap, 0)
            val textLen = recognizedTextLength(image)
            val isUtility = textLen >= UTILITY_TEXT_LEN
            result["isUtility"] = isUtility

            // 의미 분석은 문서가 아닐 때만 (영수증/지도엔 불필요)
            if (!isUtility) {
                result["hasText"] = textLen >= 20   // 메뉴판/표지판 수준
                val faces = detectFaces(image)
                result["faceCount"] = faces.first
                result["hasFace"] = faces.first > 0
                result["isSmiling"] = faces.second
                val cls = classifyImageWithLabels(image)
                result["isFood"] = cls.food
                result["isLandscape"] = cls.landscape
                result["isLandmark"] = cls.landmark
                result["sceneLabels"] = cls.topLabels
            }
        } catch (e: Exception) {
            result["error"] = e.message ?: "ANALYZE_ERROR"
        } finally {
            bitmap.recycle()
        }
        return result
    }

    // ─── 파일 경로 ───
    private fun filePath(uri: String): String? = when {
        uri.startsWith("file://") -> Uri.parse(uri).path
        uri.startsWith("content://") -> null // 호출부에서 file:// 썸네일을 넘겨야 함
        else -> uri
    }

    // ─── 다운샘플 디코딩 ───
    private fun decodeDownsampled(path: String, maxPixel: Int): Bitmap? {
        if (!File(path).exists()) return null

        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeFile(path, bounds)
        val w = bounds.outWidth
        val h = bounds.outHeight
        if (w <= 0 || h <= 0) return null

        var sample = 1
        while (max(w, h) / sample > maxPixel) sample *= 2

        val opts = BitmapFactory.Options().apply {
            inSampleSize = sample
            inPreferredConfig = Bitmap.Config.ARGB_8888
        }
        return BitmapFactory.decodeFile(path, opts)
    }

    // ─── 그레이스케일 ───
    private data class Gray(val lum: IntArray, val width: Int, val height: Int)

    private fun toGray(bitmap: Bitmap, edge: Int): Gray {
        val ratio = min(1.0, edge.toDouble() / max(bitmap.width, bitmap.height))
        val w = max(1, (bitmap.width * ratio).toInt())
        val h = max(1, (bitmap.height * ratio).toInt())

        val scaled = Bitmap.createScaledBitmap(bitmap, w, h, true)
        val argb = IntArray(w * h)
        scaled.getPixels(argb, 0, w, 0, 0, w, h)
        if (scaled != bitmap) scaled.recycle()

        val lum = IntArray(w * h)
        for (i in argb.indices) {
            val c = argb[i]
            val r = (c shr 16) and 0xFF
            val g = (c shr 8) and 0xFF
            val b = c and 0xFF
            // ITU-R BT.601 가중치
            lum[i] = (r * 299 + g * 587 + b * 114) / 1000
        }
        return Gray(lum, w, h)
    }

    // ─── 노출 (평균 밝기 0~1) ───
    private fun meanLuminance(lum: IntArray): Double {
        if (lum.isEmpty()) return 0.0
        var sum = 0L
        for (p in lum) sum += p
        return (sum.toDouble() / lum.size) / 255.0
    }

    // ─── 흔들림 (라플라시안 분산) ───
    private fun laplacianVariance(lum: IntArray, w: Int, h: Int): Double {
        if (w <= 2 || h <= 2) return 0.0
        val n = (w - 2) * (h - 2)
        val resp = DoubleArray(n)
        var idx = 0
        for (y in 1 until h - 1) {
            for (x in 1 until w - 1) {
                val c = lum[y * w + x]
                val up = lum[(y - 1) * w + x]
                val down = lum[(y + 1) * w + x]
                val left = lum[y * w + (x - 1)]
                val right = lum[y * w + (x + 1)]
                resp[idx++] = (4 * c - up - down - left - right).toDouble()
            }
        }
        var mean = 0.0
        for (r in resp) mean += r
        mean /= n
        var variance = 0.0
        for (r in resp) variance += (r - mean) * (r - mean)
        return variance / n
    }

    // ─── 인식된 텍스트 길이(공백 제외) — isUtility·hasText 공용 ───
    private fun recognizedTextLength(image: InputImage): Int {
        return try {
            // 백그라운드 스레드에서 호출되므로 Tasks.await 블로킹 사용 가능
            val visionText = Tasks.await(textRecognizer.process(image))
            visionText.text.replace(Regex("\\s"), "").length
        } catch (e: Exception) { 0 }
    }

    // ─── 얼굴 수 + 웃음 → (faceCount, isSmiling) ───
    private fun detectFaces(image: InputImage): Pair<Int, Boolean> {
        return try {
            val faces = Tasks.await(faceDetector.process(image))
            val smiling = faces.any { (it.smilingProbability ?: 0f) >= SMILE_PROB }
            Pair(faces.size, smiling)
        } catch (e: Exception) { Pair(0, false) }
    }

    // ─── 분류 1회 → 카테고리 + 상위 10 라벨 ───
    private data class ClassLabels(
        val food: Boolean, val landscape: Boolean, val landmark: Boolean,
        val topLabels: List<Map<String, Any>>
    )

    private fun classifyImageWithLabels(image: InputImage): ClassLabels {
        return try {
            val all = Tasks.await(imageLabeler.process(image))
            val topLabels = all
                .filter { it.confidence >= 0.3f }
                .sortedByDescending { it.confidence }
                .take(10)
                .map { mapOf<String, Any>("label" to it.text.lowercase(), "confidence" to it.confidence.toDouble()) }

            val strong = all.filter { it.confidence >= LABEL_CONFIDENCE }.map { it.text.lowercase() }
            fun hit(keys: List<String>) = strong.any { label -> keys.any { label.contains(it) } }
            ClassLabels(hit(FOOD_KEYWORDS), hit(LANDSCAPE_KEYWORDS), hit(LANDMARK_KEYWORDS), topLabels)
        } catch (e: Exception) {
            ClassLabels(false, false, false, emptyList())
        }
    }

    // ─── 색감 통계 — 64px 축소본에서 채도/색온도/대비/어두움 ───
    private fun colorStatistics(bitmap: Bitmap): Map<String, Double> {
        val edge = 64
        val ratio = min(1.0, edge.toDouble() / max(bitmap.width, bitmap.height))
        val w = max(1, (bitmap.width * ratio).toInt())
        val h = max(1, (bitmap.height * ratio).toInt())
        val scaled = Bitmap.createScaledBitmap(bitmap, w, h, true)
        val argb = IntArray(w * h)
        scaled.getPixels(argb, 0, w, 0, 0, w, h)
        if (scaled != bitmap) scaled.recycle()

        var satSum = 0.0; var warmSum = 0.0; var lumSum = 0.0; var lumSqSum = 0.0
        var darkCount = 0
        for (c in argb) {
            val r = ((c shr 16) and 0xFF) / 255.0
            val g = ((c shr 8) and 0xFF) / 255.0
            val b = (c and 0xFF) / 255.0
            val maxC = maxOf(r, g, b); val minC = minOf(r, g, b)
            satSum += if (maxC == 0.0) 0.0 else (maxC - minC) / maxC
            warmSum += (r - b + 1.0) / 2.0
            val lum = 0.299 * r + 0.587 * g + 0.114 * b
            lumSum += lum
            lumSqSum += lum * lum
            if (lum < 0.235) darkCount++
        }
        val n = argb.size.toDouble()
        val lumMean = lumSum / n
        val variance = max(0.0, lumSqSum / n - lumMean * lumMean)
        return mapOf(
            "saturation" to satSum / n,
            "warmth" to warmSum / n,
            "contrast" to min(1.0, Math.sqrt(variance) * 4.0),
            "darkness" to darkCount / n,
        )
    }

    // ─── dHash — 9x8 그레이스케일 가로 인접 비교, 64bit 16진수 16자 (iOS와 동일 알고리즘) ───
    private fun differenceHash(bitmap: Bitmap): String {
        val w = 9; val h = 8
        val scaled = Bitmap.createScaledBitmap(bitmap, w, h, true)
        val argb = IntArray(w * h)
        scaled.getPixels(argb, 0, w, 0, 0, w, h)
        if (scaled != bitmap) scaled.recycle()

        val lum = IntArray(w * h)
        for (i in argb.indices) {
            val c = argb[i]
            lum[i] = (((c shr 16) and 0xFF) * 299 + ((c shr 8) and 0xFF) * 587 + (c and 0xFF) * 114) / 1000
        }
        var bits = 0L
        for (y in 0 until h) {
            for (x in 0 until w - 1) {
                bits = bits shl 1
                if (lum[y * w + x] > lum[y * w + x + 1]) bits = bits or 1L
            }
        }
        return String.format("%016x", bits)
    }
}
